File: config/core/schemas/disease_detection.py
"""
KisanAI OS
Disease Detection Schemas
Version: 1.0.0

Future-compatible prediction result. A real model will populate
``disease_name``, ``confidence`` and ``model``; until then these stay
``None`` with ``status = MODEL_NOT_CONFIGURED`` so "no model" is never
confused with "healthy" or "disease detected". No local filesystem paths
are ever exposed.
"""

from typing import Optional

from pydantic import BaseModel


class DiseaseDetectionOut(BaseModel):
    """Structured disease-detection response."""

    success: bool = True
    status: str
    crop: Optional[str] = None
    disease_name: Optional[str] = None
    confidence: Optional[float] = None
    model: Optional[str] = None
    message: Optional[str] = ""

File: mobile/lib/models/disease_detection.dart
/// Structured disease-detection response (`DiseaseDetectionOut`).
///
/// Status values mirror the backend provider layer:
/// - `HEALTHY` — no disease detected
/// - `DISEASE_DETECTED` — a disease was found
/// - `MODEL_NOT_CONFIGURED` — no model is bundled; never confused with
///   "healthy" or a real diagnosis.
class DiseaseDetectionResult {
  const DiseaseDetectionResult({
    required this.success,
    required this.status,
    required this.crop,
    required this.diseaseName,
    required this.confidence,
    required this.model,
    required this.message,
  });

  static const String statusHealthy = 'HEALTHY';
  static const String statusDiseaseDetected = 'DISEASE_DETECTED';
  static const String statusModelNotConfigured = 'MODEL_NOT_CONFIGURED';

  final bool success;
  final String status;
  final String? crop;
  final String? diseaseName;
  final double? confidence;
  final String? model;
  final String? message;

  bool get isHealthy => status == statusHealthy;
  bool get isDiseaseDetected => status == statusDiseaseDetected;
  bool get isModelNotConfigured => status == statusModelNotConfigured;

  factory DiseaseDetectionResult.fromJson(Object? json) {
    final map = json as Map<String, dynamic>;
    return DiseaseDetectionResult(
      success: map['success'] as bool? ?? true,
      status: map['status'] as String? ?? statusModelNotConfigured,
      crop: map['crop'] as String?,
      diseaseName: map['disease_name'] as String?,
      confidence: (map['confidence'] as num?)?.toDouble(),
      model: map['model'] as String?,
      message: map['message'] as String?,
    );
  }
}
